### PR TITLE
Implement SPIFFE TLS listening

### LINKED ISSUE #73: Implement SPIFFE TLS listening
The `spiffetls` package is used to facilitate SPIFFE-authenticated TLS connectivity between peers leveraging the Workload API. It essentially ties together the configuration produced by the `tlsconfig` package with the sources available in the `workloadapi` package.

This issue is to implement the listening functionality, as scaffolded in [`listen.go`](https://github.com/spiffe/go-spiffe/blob/v2-api/v2/spiffetls/listen.go) and [`option.go`](https://github.com/spiffe/go-spiffe/blob/v2-api/v2/spiffetls/option.go) (the listening related options, that is).

This issue depends on #64 and #69.

### PR DESCRIPTION
**Pull request by amartinezfayo**
- Implement SPIFFE TLS listening

Tests will be added in a separate PR.
Fixes #73

### CODE REVIEW

File: v2/spiffetls/listen.go [hunk 1]
```diff
@@ -24,20 +26,78 @@ func Listen(ctx context.Context, network, laddr string, authorizer tlsconfig.Aut
...
 // controls the authentication and authorization exercised during the TLS
 // handshake.
-func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (_ net.Listener, err error) {
+	m := mode.get()
+	source := m.source
+	if source == nil {
+		source, err = workloadapi.NewX509Source(ctx, m.options...)
+		// Close source if there is a failiure after this point
```
Comment (azdagron) [inline v2/spiffetls/listen.go:59]: nit: `failure`

File: v2/spiffetls/listen.go [hunk 2]
```diff
@@ -24,20 +26,78 @@ func Listen(ctx context.Context, network, laddr string, authorizer tlsconfig.Aut
...
+		tlsCert, err := tlsconfig.GetTLSCertificate(m.svid)
+		if err != nil {
+			return nil, err
+		}
+		tlsconfig.HookMTLSWebServerConfig(tlsConfig, tlsCert, m.bundle, m.authorizer)
+	default:
+		return nil, spiffetlsErr.New("unknown TLS hook type: %v", m.tlsType)
+	}
+
+	return tls.NewListener(inner, tlsconfig.MTLSServerConfig(m.source, m.bundle, m.authorizer)), nil
```
Comment (azdagron) [inline v2/spiffetls/listen.go:102]:
```suggestion
	return tls.NewListener(inner, tlsConfig), nil
```

File: v2/spiffetls/listen.go [hunk 3]
```diff
@@ -24,20 +26,78 @@ func Listen(ctx context.Context, network, laddr string, authorizer tlsconfig.Aut
...
+		tlsConfig = opt.baseTLSConf
+	}
+
+	switch m.tlsType {
+	case typeTLSServer:
+		tlsconfig.HookTLSServerConfig(tlsConfig, m.svid)
+	case typeMTLSServer:
+		tlsconfig.HookMTLSServerConfig(tlsConfig, m.svid, m.bundle, m.authorizer)
+	case typeMTLSWebServer:
+		tlsCert, err := tlsconfig.GetTLSCertificate(m.svid)
```
Comment (azdagron) [inline v2/spiffetls/listen.go:93]: The web server does not serve an X509-SVID and should not get the certificate from the source. It serves the certificate passed into the various `MTLSWebServer*` modes, i.e. `m.cert`.

File: v2/spiffetls/listen.go [hunk 4]
```diff
@@ -24,20 +26,74 @@ func Listen(ctx context.Context, network, laddr string, authorizer tlsconfig.Aut
...
-func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (_ net.Listener, err error) {
+	m := mode.get()
+	source := m.source
+	if source == nil {
+		source, err = workloadapi.NewX509Source(ctx, m.options...)
+		// Close source if there is a failure after this point
+		defer func() {
+			if err != nil {
```
Comment (marcosy) [inline v2/spiffetls/listen.go:61]: While writing unit tests I realized that we should check for `source != nil` also.

```suggestion
			if err != nil && source != nil {
```

If you agree, would you mind update it in dial as well?

**Pull request merged**